### PR TITLE
Rdf Collapsing Intro

### LINKED ISSUE #42: Investigate SPARQL Queries for collapsing nodes during RDF to networkx transformation
We are interested in using networkx to perform complex graph algorithms, and already use it to build visualizations. As RDF represents all data as nodes, we will want to collapse down DataProperties onto the nodes that are created in the networkx graph. 

This is the first ticket towards that capability that investigates the SPARQL queries needed to perform the collapsing process.

### PR DESCRIPTION
**Pull request by lnijhawan**
to run:
- doitlab
- run RDF_to_LPG_Example.ipynb

to run tests:
- pytests examples/tests

CHANGES:
- changed from `rdflib_to_networkx_graph` to `rdflib_to_networkx_digraph`
- more detailed description and specific examples of node collapsing via the new `collapse_predicates` function
- added a `PredicateMultiselectApp` which uses a combination of widgets to give a user the ability to select predicates they wish to collapse in an RDF graph
- developed `CustomURIRef` class, a way to store a URIRef URI but receive a 'pretty'/truncated version of the name for display purposes
- added pytests for both `CustomURIRef` class and the `collapse_predicates` function

[Binder](https://mybinder.org/v2/gh/lnijhawan/ipyradiant/rdf_collapsing?filepath=%2Fexamples%2FLiteralObjects_Example.ipynb)

### CODE REVIEW

File: Untitled.ipynb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (nrbgt) [inline Untitled.ipynb]: please remove. should already have `Untitled.*` in `.gitignore`, if not add.

File: examples/Cytoscape_Example.ipynb [hunk 2]
```diff
@@ -1,74 +1,74 @@
-{
```
Comment (nrbgt) [inline examples/Cytoscape_Example.ipynb:1]: all this pain suggests your repo is doing line ending trash... highly recommend running this in git bash:

```bash
git config --global core.autocrlf true
```

and then re-checking out the repo and running the linter.

File: src/ipyradiant/visualization/cytoscape.py [hunk 3]
```diff
@@ -76,9 +80,16 @@ def __init__(self, *args, **kwargs):
             (self.click_output_box.layout, "visibility"),
             lambda x: "visible" if x else "hidden",
         )
-        self.children = [
-            self.cyto_widget,
-        ]
+
+        if self.show_outputs == True:
```
Comment (nrbgt) [inline src/ipyradiant/visualization/cytoscape.py]: Linter sez:
```
cytoscape.py:84:30: E712 comparison to True should be 'if cond is True:' or 'if cond:'
```

i concur

File: Untitled.ipynb (hunk 1 above)
Comment (lnijhawan) [reply Untitled.ipynb]: Whoops. not sure how that got inthere.

File: examples/tests/test_PredCollapsing.py [hunk 4]
```diff
@@ -67,7 +67,7 @@ def test_node_data(test_graph):
         collapsed_netx.nodes.get(URIRef("www.testing.com/Luke")).get(
             rdflib.term.URIRef("www.testing.com/has")
         )
-        == "scarves"
+        == "one rother"
```
Comment (zwelz3) [inline examples/tests/test_PredCollapsing.py]: "one brother"?

File: examples/tests/test_PredCollapsing.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+def test_correct_length(test_graph):
+    preds_to_collapse = [
+        rdflib.URIRef("www.testing.com/has"),
+        rdflib.URIRef("www.testing.com/goes"),
+    ]
+    subjects = [
+        rdflib.URIRef("www.testing.com/EmiratesStadium"),
+    ]
+    netx_version = rdflib_to_networkx_graph(test_graph)
```
Comment (zwelz3) [inline examples/tests/test_PredCollapsing.py]: Did you decide on using this function in the end? Can you explain why you decided to use this function and/or how you avoided the issues we previously discussed?

File: src/ipyradiant/basic_tools/custom_uri_ref.py [hunk 6]
```diff
@@ -7,18 +7,23 @@
...
+    maintaining access to the full URIRef via the 'uri' object property.
+
+    params:
+    :uri: URIRef, the URIRef one wishes to truncate
+    :namespaces: either a dict or NamespaceManager, this corresponds to the set of namespaces the URI should be
+    truncated according to.
+    """
+
     # TODO: investigate using rdflib.URIRef as a baseclass
     # TODO: investigate better way to handle namespace collapsing via __repr__
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/custom_uri_ref.py]: I think this TODO is done and can be removed. 

Why did you add a `__hash__` method?

File: src/ipyradiant/basic_tools/object_literal_collapsing.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    def make_default_graph(self):
+        return Graph()
+
+    @T.default("lit_button")
+    def make_default_button(self):
+
+        layout = W.Layout(width="300px", height="40px")  # set width and height
+
+        button = W.Button(
+            description="Add predicates where object is a literal",
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/object_literal_collapsing.py:62]: The behavior of this button is getting close to complete. I noticed that adding a predicate (where the object is not a Literal) and then clicking the button removes the predicate that I added. We will want to fix that, but other than that it looks good. 

Let me know if you need additional clarification.

File: examples/LiteralObjects_Example.ipynb [hunk 8]
```diff
@@ -0,0 +1,240 @@
+{
```
Comment (zwelz3) [inline examples/LiteralObjects_Example.ipynb]: Let's have a brief discussion about the motivation behind this notebook when you go through the review.

Comment (zwelz3) [reply examples/LiteralObjects_Example.ipynb]: I think we can clean up the discussion a little.

File: src/ipyradiant/basic_tools/custom_uri_ref.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2020 ipyradiant contributors.
+# Distributed under the terms of the Modified BSD License.
+from typing import Union
+
+from rdflib import Graph, URIRef
+from rdflib.namespace import NamespaceManager
+
+
+class CustomURIRef:
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/custom_uri_ref.py:9]: I think we may want to use this as the key for nodes in the collapsed graph so that they show up nice and pretty as well.

**Review by zwelz3 — CHANGES_REQUESTED** [review]
Can you update your PR description so that we know how to review, and what we should be looking at? 

I noticed you added some callbacks to datashader, but they are not discussed and not in the scope of the PR. It is okay to add, but you should motivate the decision with discussion.

 Let's hold a short meeting once you have time to review the comments here. I think we can wrap up this PR relatively soon.

File: src/ipyradiant/basic_tools/custom_uri_ref.py (hunk 6 above)
Comment (lnijhawan) [reply src/ipyradiant/basic_tools/custom_uri_ref.py]: added `__hash__` so that it can be key of dictionary.

File: examples/tests/test_PredCollapsing.py (hunk 5 above)
Comment (lnijhawan) [reply examples/tests/test_PredCollapsing.py]: ended up staying with this one because:
1) didnt see a huge difference or gain of information in this one vs. di or multidi graph,
2) with this one all the edges are still in tact so i dont think there is any information lost, and 
3) this is what we currently use for datashader /cytoscape vis so i thought it might be best to stay uniform.

that being said it might be a productive discussion to have or at least something to keep in mind.

File: src/ipyradiant/basic_tools/object_literal_collapsing.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        self.multiselect.available_things_list = truncated_available_uris
+        self.multiselect.selected_things_list = truncated_selected_uris
+
+
+def collapse_preds(
+    netx_graph: networkx.Graph,
+    preds_to_collapse: list,
+    subjects: list,
+    namespaces: Union[dict, NamespaceManager],
+):
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/object_literal_collapsing.py]: `) -> networkx.Graph:`

File: design/architecture/architecture.dio.svg [hunk 11]
```diff
@@ -1 +1 @@
...
...height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 368px; height: 1px; padding-top: 1060px; margin-left: 71px;"><div style="box-sizing: border-box; font-size: 0; text-align: center; "><div style="display: inline-block; font-size: 12px; font-family: Helvetica; color: #000000; line-height: 1.2; pointer-events: all; white-space: normal; word-wrap: normal; ">NOTE: in the future, we will want to implement lots of different types of filters for predicates, not just when object is a literal.</div></div></div></foreignObject><text x="255" y="1064" fill="#000000" font-family="Helvetica" font-size="12px" text-anchor="middle">NOTE: in the future, we will want to implement lots of differ...</text></switch></g></g></g><switch><g requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"/><a transform="translate(0,-5)" xlink:href="https://desk.draw.io/support/solutions/articles/16000042487" target="_blank"><text text-anchor="middle" font-size="10px" x="50%" y="100%">Viewer does not support full SVG 1.1</text></a></switch></svg>
```
Comment (zwelz3) [inline design/architecture/architecture.dio.svg]: Lets remove this change. We will add updated architecture elements when we have a more complete RDF->LPG workflow.

File: examples/Cytoscape_Example.ipynb (hunk 2 above)
Comment (zwelz3) [reply examples/Cytoscape_Example.ipynb:1]: Bumping this comment for @lukasnijhawan to address.

File: examples/RDF_to_LPG_Example.ipynb [hunk 12]
```diff
@@ -0,0 +1,266 @@
+{
```
Comment (zwelz3) [inline examples/RDF_to_LPG_Example.ipynb:1]: This file isn't in the index.ipynb which links all the examples. Lets add a section to the index called `Workflow Examples` and add this notebook with the label `Collapsing RDF to a Latent Property Graph (networkx)`.

File: examples/RDF_to_LPG_Example.ipynb [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    "disappear."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "example_1 = collapse_predicates(\n",
```
Comment (zwelz3) [inline examples/RDF_to_LPG_Example.ipynb:190]: This example needs a rework. The node that is checked should be the object not the predicate.

File: examples/Test_MultiPanelSelect_Widget.ipynb [hunk 14]
```diff
@@ -1,147 +1,147 @@
...
-   "name": "python",
-   "nbconvert_exporter": "python",
-   "pygments_lexer": "ipython3",
-   "version": "3.7.8"
-  }
- },
- "nbformat": 4,
- "nbformat_minor": 4
-}
+{
```
Comment (zwelz3) [inline examples/Test_MultiPanelSelect_Widget.ipynb:1]: Once line endings are patched, I think this diff should dissappear.

File: src/ipyradiant/basic_tools/custom_uri_ref.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+    def __repr__(self):
+        if type(self.namespaces) == dict:
+            g = Graph()
+            for key, value in self.namespaces.items():
+                g.namespace_manager.bind(key, value)
+            return URIRef(self.uri).n3(g.namespace_manager)
+        return URIRef(self.uri).n3(self.namespaces)
+
+    # todo: can we support other = URIRef?
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/custom_uri_ref.py:37]: Can you elaborate what you mean by this comment (be more descriptive in the code file)?

File: src/ipyradiant/basic_tools/object_literal_collapsing.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+        This difference is then converted into what should go to each side of the multiselect, and subsequently
+        passed to the CustomURIRef class in order to get a 'pretty' representation of the URIRef.
+        """
+        # query for all predicates where object is not literal
+        predicates_to_uris = """
+            SELECT DISTINCT  ?p
+            WHERE {
+                    ?s ?p ?o .
+                FILTER (!isLiteral(?o))
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/object_literal_collapsing.py:101]: all commands (incl. `FILTER`) should be indented to the same level as the query. No need to update this right now, but just a message for the future.

File: src/ipyradiant/basic_tools/object_literal_collapsing.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+        for _, p, _ in netx_graph[s][o]["triples"]:
+
+            if not isinstance(p, URIRef):
+                raise ValueError(f"Predicate must be URIRef not a '{type(p)}'.")
+            if p not in predicates_to_collapse:
+                continue
+
+            # add data to subject node
+            # TODO: custom representation?
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/object_literal_collapsing.py]: Can you expand on what you mean by this TODO (in the code file).

File: src/ipyradiant/basic_tools/selection_widget.py [hunk 18]
```diff
@@ -42,12 +67,33 @@ def _make_selected_things(self):
     @trt.default("add_button")
     def _make_add_button(self):
         return ipyw.Button(
-            description="Add",
+            description="Add -->".center(12),
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/selection_widget.py:70]: This magic number may cause issues if the buttons are made to by dynamically sized, but we will not worry about this for now. Just be aware that this will cause weird styling in some edge cases.

File: src/ipyradiant/basic_tools/selection_widget.py [hunk 19]
```diff
@@ -42,12 +67,33 @@ def _make_selected_things(self):
...
+        self.available_things = self._make_available_things()
+
+    @trt.observe("selected_things_list")
+    def _change_selected_items(self, change):
+        self.selected_things = self._make_selected_things()
+
+    @trt.observe(
+        "selected_things", "available_things", "right_panel_text", "left_panel_text"
+    )
+    def _change_third_column(self, change):
```
Comment (zwelz3) [inline src/ipyradiant/basic_tools/selection_widget.py]: right panel, left panel, third column

The naming is a bit confusing. It is not obvious to me what this method is doing based on the name and the code. Maybe a docstring for the method and/or rename the function?

File: src/ipyradiant/basic_tools/object_literal_collapsing.py (hunk 17 above)
Comment (lnijhawan) [reply src/ipyradiant/basic_tools/object_literal_collapsing.py]: actually thats already taken care of, forgot to get rid of it. my bad

**Comment by zwelz3** [comment]
In the PR description, I would make sure we are precise in the test procedure. 
Need to `conda activate envs/develop`, and the command is `pytest`. 
Everything else looks good. We can circle back on other improvements on the next ticket.
